parser.go

```diff
@@ -40,6 +40,57 @@ type Node[V comparable] struct {
 	Column int
 }
 
+// Left returns the left child in the case of a binary tree.
+func (p *Node[V]) Left() *Node[V] {
+	if len(p.Children) > 0 {
+		return p.Children[0]
+	}
+	return nil
+}
+
+// SetLeft sets the left child in the case of a binary tree and returns the
+// previous value.
+func (p *Node[V]) SetLeft(l *Node[V]) *Node[V] {
+	for len(p.Children) < 1 {
+		p.Children = append(p.Children, nil)
+	}
+	old := p.Children[0]
+	p.Children[0] = l
+	l.Parent = p
+	return old
+}
+
+// Right returns the right child in the case of a binary tree.
+func (p *Node[V]) Right() *Node[V] {
+	if len(p.Children) > 1 {
+		return p.Children[1]
+	}
+	return nil
+}
+
+// SetRight sets the right child in the case of a binary tree and returns the
+// previous value.
+func (p *Node[V]) SetRight(r *Node[V]) *Node[V] {
+	for len(p.Children) < 2 {
+		p.Children = append(p.Children, nil)
+	}
+	old := p.Children[1]
+	p.Children[1] = r
+	r.Parent = p
+	return old
+}
+
+// ReplaceChild replaces the first given node with another node. The inserted
+// node's parent is updated. The removed node's parent is not updated.
+func (p *Node[V]) ReplaceChild(l, r *Node[V]) {
+	for i := range p.Children {
+		if p.Children[i] == l {
+			p.Children[i] = r
+			r.Parent = p
+		}
```

Comment:
Will there be trees with the same node multiple times?  There could be a `return` after finding the first match.  Also, probably a node will not have many children and any expected performance difference won't matter.

Reply:
It's possible since the tree is largely constructed by the user, though it wouldn't be a proper tree since it would create an [undirected cycle](https://en.wikipedia.org/wiki/Tree_(data_structure)#Examples_of_trees_and_non-trees).

I think returning early is fine.